sa/sa.go

```diff
@@ -1375,11 +1378,50 @@ func (ssa *SQLStorageAuthority) NewAuthorizations2(ctx context.Context, req *sap
 		if err != nil {
 			return nil, err
 		}
-		err = ssa.dbMap.Insert(am)
+
+		// Each authz needs a (?,?...), in the VALUES block. We need one
+		// for each element in the authzFields string with no trailing ','.
+		fmt.Fprintf(&questionsBuf, "(%s),", strings.TrimRight(strings.Repeat(" ?,", 11), ","))
+
+		// The query arguments must follow the order of the authzFields string.
+		// Note that the AttemptedAt field is not included in the authzFields.
+		queryArgs = append(queryArgs,
+			am.ID,
+			am.IdentifierType,
+			am.IdentifierValue,
+			am.RegistrationID,
+			am.Status,
+			am.Expires,
+			am.Challenges,
+			am.Attempted,
+			am.Token,
+			am.ValidationError,
+			am.ValidationRecord,
+		)
+	}
+
+	// At this point, the VALUES block question-string has a trailing comma, we need
+	// to remove it to make sure we're valid SQL.
+	questionsTrimmed := strings.TrimRight(questionsBuf.String(), ",")
+	query := fmt.Sprintf("INSERT INTO authz2 (%s) VALUES %s RETURNING id;", authzFields, questionsTrimmed)
+
+	rows, err := ssa.dbMap.Db.QueryContext(ctx, query, queryArgs...)
+	if err != nil {
+		return nil, err
+	}
+	for rows.Next() {
+		var idField int64
+		err = rows.Scan(&idField)
 		if err != nil {
 			return nil, err
```

Comment:
A rows.Close() should happen right before returning out of here - the error from rows.Close() can probably be thrown out too.

Reply:
my poor defer

Reply:
> my poor defer

💔